### PR TITLE
[Model Monitoring] Fix monitoring feature set update

### PR DESCRIPTION
**Pull request by davesh0812**
[ML-5981](https://iguazio.atlassian.net/browse/ML-5981)

[ML-5981]: https://iguazio.atlassian.net/browse/ML-5981?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/system/model_monitoring/test_app.py [hunk 1]
```diff
@@ -137,9 +139,35 @@ def _test_tsdb_record(cls, ep_id: str) -> None:
...
+        )
+        assert is_metadata_saved, "Dataframe does not contains the metadata columns"
+
+    @classmethod
+    def _test_v3io_records(
+        cls, ep_id: str, inputs: set[str], outputs: set[str]
+    ) -> None:
         cls._test_kv_record(ep_id)
         cls._test_tsdb_record(ep_id)
+        cls._test_apps_parquet(ep_id, inputs, outputs)
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]: Move this to the first place before the KV + TSDB tests. If there are no parquets, there will also be no data

File: tests/system/model_monitoring/test_app.py [hunk 2]
```diff
@@ -137,9 +139,35 @@ def _test_tsdb_record(cls, ep_id: str) -> None:
...
+                mlrun.get_or_create_project(cls.project_name),
+                kind=mm_constants.FileTargetKind.APPS_PARQUET,
+            )
+        )
+        df = ParquetTarget(
+            path=f"{parquet_apps_directory}/key={ep_id}",
+        ).as_df()
+
+        is_inputs_saved = inputs.issubset(df.columns)
+        assert is_inputs_saved, "Dataframe does not contains the input columns"
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]:
```suggestion
        assert is_inputs_saved, "Dataframe does not contain the input columns"
```

File: tests/system/model_monitoring/test_app.py [hunk 3]
```diff
@@ -137,9 +139,35 @@ def _test_tsdb_record(cls, ep_id: str) -> None:
...
+        is_output_saved = outputs.issubset(df.columns)
+        assert is_output_saved, "Dataframe does not contains the output columns"
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]:
```suggestion
        assert is_output_saved, "Dataframe does not contain the output columns"
```

File: tests/system/model_monitoring/test_app.py [hunk 4]
```diff
@@ -82,6 +82,7 @@ def __post_init__(self) -> None:
 
 
 class _V3IORecordsChecker:
+    project_name = None
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]:
```suggestion
    project_name: str
```

File: tests/system/model_monitoring/test_app.py [hunk 5]
```diff
@@ -93,6 +94,7 @@ def custom_setup_class(cls, project_name: str) -> None:
         cls._tsdb_storage = ModelMonitoringWriter._get_v3io_frames_client(
             cls._v3io_container
         )
+        cls.project_name = project_name
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]:
```suggestion
```

File: tests/system/model_monitoring/test_app.py [hunk 6]
```diff
@@ -264,35 +301,60 @@ def _infer(
...
+        dataset = load_iris()
+        inputs: set = (
+            {
+                mlrun.feature_store.api.norm_column_name(feature)
+                for feature in dataset.feature_names
+            }
+            if with_training_set
+            else {f"f{i}" for i in range(len(dataset.feature_names))}
+        )
+        outputs = {"p0"}
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]: 1. Please move this to another method. It doesn't make sense to do it here, as it's the main flow, and this is some data prep.
2. I would avoid calling `load_iris` twice in the test.

File: tests/system/model_monitoring/test_app.py [hunk 7]
```diff
@@ -492,34 +556,47 @@ def custom_setup_class(cls) -> None:
...
+                "name": "serving_7",
                 "model_name": "int_and_str_one_to_one",
                 "class_name": "OneToOne",
                 "data_point": [1, "a", 3],
                 "schema": ["f0", "f1", "f2", "p0"],
             },
         }
 
-    def _log_model(self, model_name) -> None:
+    def _log_model(self, model_name, training_set=None, label_column=None) -> None:
```
Comment (jond01) [inline tests/system/model_monitoring/test_app.py]: Nit: missing args type hints

File: tests/system/model_monitoring/assets/custom_evidently_app.py [hunk 8]
```diff
@@ -126,37 +127,37 @@ def _create_evidently_project(
         project.save()
         return project
 
-    def create_demo_project(workspace_path: str) -> tuple[Workspace, Project]:
-        workspace = Workspace.create(workspace_path)
-        project = _create_evidently_project(workspace)
-        return workspace, project
-
```
Comment (jond01) [inline tests/system/model_monitoring/assets/custom_evidently_app.py:133]: 👍

File: tests/system/model_monitoring/assets/custom_evidently_app.py [hunk 9]
```diff
@@ -126,37 +127,37 @@ def _create_evidently_project(
...
 class CustomEvidentlyMonitoringApp(EvidentlyModelMonitoringApplicationBase):
     NAME = "evidently-app-test"
 
-    def _lazy_init(self, *args, **kwargs) -> None:
-        super()._lazy_init(*args, **kwargs)
+    def __init__(
+        self,
+        evidently_workspace_path: str,
+        evidently_project_id: "STR_UUID",
+        with_training_set,
```
Comment (jond01) [inline tests/system/model_monitoring/assets/custom_evidently_app.py]: 1. Missing type hint.
2. Is it even necessary? Are there current usages with `with_training_set=False`?

File: tests/system/model_monitoring/assets/custom_evidently_app.py [hunk 10]
```diff
@@ -126,37 +127,37 @@ def _create_evidently_project(
...
+            self.columns = [f"f{i}" for i in range(4)]
         self.train_set = pd.DataFrame(iris.data, columns=self.columns)
 
     def _init_evidently_project(self) -> None:
         if self.evidently_project is None:
             if isinstance(self.evidently_project_id, str):
                 self.evidently_project_id = UUID(self.evidently_project_id)
-            self.context.logger.info(
-                "Creating evidently project", id=self.evidently_project_id
-            )
```
Comment (jond01) [inline tests/system/model_monitoring/assets/custom_evidently_app.py:159]: Why? This log is helpful

File: mlrun/model_monitoring/stream_processing.py [hunk 11]
```diff
@@ -1225,3 +1237,19 @@ def get_endpoint_record(project: str, endpoint_id: str):
         project=project,
     )
     return model_endpoint_store.get_model_endpoint(endpoint_id=endpoint_id)
+
+
+def update_monitoring_feature_set(
+    endpoint_record: dict[str, typing.Any], feature_names: list, feature_values: list
```
Comment (jond01) [inline mlrun/model_monitoring/stream_processing.py]: ?
```suggestion
    endpoint_record: dict[str, typing.Any], feature_names: list[str], feature_values: list[typing.Any]
```

**Review by jond01 — COMMENTED** [review]
Nice, I had some more comments

File: tests/system/model_monitoring/test_app.py (hunk 6 above)
Comment (davesh0812) [reply tests/system/model_monitoring/test_app.py]: I moved it to  the `_log_model()` method

File: tests/system/model_monitoring/assets/custom_evidently_app.py (hunk 9 above)
Comment (davesh0812) [reply tests/system/model_monitoring/assets/custom_evidently_app.py]: It does use it inside `_init_iris_data()` method + added type hint

File: tests/system/model_monitoring/assets/custom_evidently_app.py (hunk 10 above)
Comment (davesh0812) [reply tests/system/model_monitoring/assets/custom_evidently_app.py:159]: I moved your implementation to the `_init_()` method, and while we are in the init we don't have a `self.context`

File: tests/system/model_monitoring/assets/custom_evidently_app.py (hunk 9 above)
Comment (jond01) [reply tests/system/model_monitoring/assets/custom_evidently_app.py]: Regarding 2 - I don't see any usage of this parameter with `False`.

Edit: ok, it was set indirectly through parametrization.

**Review by jond01 — APPROVED** [review]

**Pull request merged**